Update HTTP/3 support for HttpClient in .NET 7 (#31554)
dotnet/docs: docs/core/extensions/httpclient-http3.md

---
title: Use HTTP/3 with HttpClient

> description: Learn how to use the HttpClient to access HTTP/3 servers in .NET 6 and .NET 7

Reviewer: ```suggestion
description: Learn how to use the HttpClient to access HTTP/3 servers in .NET
```

Remove the version from the description.

- Faster response time of the first request. QUIC and HTTP/3 negotiate the connection in fewer round-trips between the client and the server. The first request reaches the server faster.
- Improved experience when there is connection packet loss. HTTP/2 multiplexes multiple requests via one TCP connection. Packet loss on the connection affects all requests. This problem is called "head-of-line blocking". Because QUIC provides native multiplexing, lost packets only impact the requests where data has been lost.

> - Supports transitioning between networks. This feature is useful for mobile devices where it is common to switch between WIFI and cellular networks as a mobile device changes location. Currently, HTTP/1.1 and HTTP/2 connections fail with an error when switching networks. An app or web browser must retry any failed HTTP requests. HTTP/3 allows the app or web browser to seamlessly continue when a network changes. HttpClient and Kestrel do not support network transitions in .NET 7. It may be available in a future release.

Reviewer: ```suggestion
- Supports transitioning between networks. This feature is useful for mobile devices where it is common to switch between WIFI and cellular networks as a mobile device changes location. Currently, HTTP/1.1 and HTTP/2 connections fail with an error when switching networks. An app or web browser must retry any failed HTTP requests. HTTP/3 allows the app or web browser to seamlessly continue when a network changes. `HttpClient` and Kestrel do not support network transitions in .NET 7. It may be available in a future release.
```

> [!IMPORTANT]
>

> > HTTP/3 is fully supported since .NET 7 only. In .NET 6, HTTP/3 is available as a _preview feature_ because the HTTP/3 specification was not yet finalized and behavioral or performance issues may exist in HTTP/3 with .NET 6.

Reviewer 2: ```suggestion
> HTTP/3 is fully supported since .NET 7 only. In .NET 6, HTTP/3 is available as a _preview feature_ because the HTTP/3 specification was not yet finalized and behavioral or performance problems may exist in HTTP/3 with .NET 6.
```

### Linux

- OpenSSL 1.1

> On Linux, libmsquic is published via Microsoft's official Linux package repository packages.microsoft.com. To consume it, it must be added manually. See [Linux Software Repository for Microsoft Products](/windows-server/administration/linux-package-repository-for-microsoft-software). After configuring the package feed, it can be installed via the package manager of your distro, for example, for Ubuntu:

Reviewer: ```suggestion
On Linux, libmsquic is published via Microsoft's official Linux package repository <packages.microsoft.com>. To consume it, it must be added manually. For more information, see [Linux Software Repository for Microsoft Products](/windows-server/administration/linux-package-repository-for-microsoft-software). After configuring the package feed, it's installed via the package manager of your distro, for example, for Ubuntu:
```

```bash
sudo apt install libmsquic
```

> [!NOTE]
> .NET 7 is only compatible with 2.1+ versions of libmsquic.

### macOS

HTTP/3 is not currently supported on macOS but may be available in a future release.

## Using HttpClient

The following code example uses [top-level statements](../../csharp/fundamentals/program-structure/top-level-statements.md) and demonstrates how to specify HTTP3 in the request:

:::code language="csharp" source="snippets/http/http3/Program.cs":::

## HTTP/3 Support in .NET 6

> In .NET 6, HTTP/3 is available as a _preview feature_ because the HTTP/3 specification was not yet finalized and behavioral or performance problems may exist in HTTP/3 with .NET 6. For more information on preview features, see [**the preview features specification**](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md#are-preview-features-supported).

Reviewer: ```suggestion
In .NET 6, HTTP/3 is available as a _preview feature_ because the HTTP/3 specification was not yet finalized. Behavioral or performance problems may exist in HTTP/3 with .NET 6. For more information on preview features, see [the preview features specification](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md#are-preview-features-supported).
```

To enable HTTP/3 support in .NET 6, include the following in the project file to enable HTTP/3 with HttpClient:

```xml
<ItemGroup>
    <RuntimeHostConfigurationOption Include="System.Net.SocketsHttpHandler.Http3Support" Value="true" />
</ItemGroup>
```

> Alternatively, you can use [`AppContext.SetSwitch`](/dotnet/api/system.appcontext.setswitch).

Reviewer: ```suggestion
Alternatively, you can call <xref:System.AppContext.SetSwitch%2A?displayProperty=fullName> from your app code, or set the `DOTNET_SYSTEM_NET_HTTP_SOCKETSHTTPHANDLER_HTTP3SUPPORT` environment variable to `true`.
```
